### PR TITLE
[timeseries] Add horizon_weight support for TimeSeriesPredictor

### PR DESCRIPTION
**Pull request by shchur**
*Issue #, if available:*

*Description of changes:* 
- Second attempt at #5058, this time folding `horizon_weight` and `eval_metric_seasonal_period` into properties of the `TimeSeriesScorer`
- Add an optional argument `horizon_weight: list[float] | None` to the `TimeSeriesPredictor` and all forecasting metrics that allows assigning custom weights to each time step in the forecast horizon when computing the metric. 
- When computing all metrics, we create an array of raw error values of shape `[num_items, prediction_length]`. We then multiply this array with `horizon_weight.reshape[1, prediction_length]`, before applying the final aggregation step. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: timeseries/src/autogluon/timeseries/predictor.py [hunk 1]
```diff
@@ -1464,7 +1474,7 @@ def select_target(ts_df: TimeSeriesDataFrame) -> TimeSeriesDataFrame:
             target=self.target,
             prediction_length=self.prediction_length,
             eval_metric=self.eval_metric.name,
-            eval_metric_seasonal_period=self.eval_metric_seasonal_period,
+            eval_metric_seasonal_period=self.eval_metric.seasonal_period,
```
Comment (canerturkmen) [inline timeseries/src/autogluon/timeseries/predictor.py:1477]: why do we need `seasonal_period` in simulation artifacts? for bookkeeping and consistency only? if so, should horizon weight also be here?

**Review by canerturkmen — COMMENTED** [review]
Thanks a lot for the great PR, I think this is headed in a great direction. Left some comments here and there.

File: timeseries/src/autogluon/timeseries/metrics/point.py [hunk 2]
```diff
@@ -38,10 +38,20 @@ class RMSE(TimeSeriesScorer):
...
+        data_future: TimeSeriesDataFrame,
+        predictions: TimeSeriesDataFrame,
+        target: str = "target",
+        prediction_length: int = 1,
+        horizon_weight: Optional[np.ndarray] = None,
+        **kwargs,
     ) -> float:
         y_true, y_pred = self._get_point_forecast_score_inputs(data_future, predictions, target=target)
-        return np.sqrt(self._safemean((y_true - y_pred) ** 2))
+        y_true, y_pred = y_true.to_numpy(), y_pred.to_numpy()
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/metrics/point.py:50]: Yes, we have tests checking equivalence to GluonTS for all metrics (+pinned scores for metrics absent from GluonTS like RMSSE).

File: timeseries/src/autogluon/timeseries/predictor.py [hunk 3]
```diff
@@ -93,6 +93,14 @@ class TimeSeriesPredictor:
     eval_metric_seasonal_period : int, optional
         Seasonal period used to compute some evaluation metrics such as mean absolute scaled error (MASE). Defaults to
         ``None``, in which case the seasonal period is computed based on the data frequency.
+    horizon_weight : List[float], optional
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/predictor.py:96]: This is mostly about efficiency and numerical stability rather than being intuitive. We always have a matrix `errors` of shape `[num_items, prediction_length]`. There are two ways to apply different weights per item / per time step.

## Current approach (no weights = all weights are equal to 1)
If we have an array `horizon_weight` of shape `[1, prediction_length]` where values sum up to `prediction_length`, and an array `item_weight` of shape `[num_items, 1]`, where values sum up to `num_items`, the averaging logic is quite simple
```python
errors: np.ndarray,  # shape [num_items, prediction_length]
if horizon_weight is not None:
    errors *= horizon_weight
if item_weight is not None:
    errors *= item_weight
return np.nanmean(errors)
```

## Alternative approach (no weights = all weights are equal to 1/prediction_length or 1/num_items)
If instead `horizon_weight` and `item_weight` each summed up to 1, the code would be a lot less elegant. We won't be able to just optionally multiple the errors with some weights, but rather we will need to perform reduction after each multiplication. We won't be able to use [`np.average`](https://numpy.org/doc/stable/reference/generated/numpy.average.html) for this reduction since it doesn't support NaN values, so we will need to us `np.nansum`. So the code will look something like
```python
errors: np.ndarray,  # shape [num_items, prediction_length]
if horizon_weight is not None:
    errors = np.nansum(errors * horizon_weight, axis=1)
else:
    errors = np.nanmean(errors, axis=1)
# now `errors` has shape [num_items]
if item_weight is not None:
    return np.nansum(errors * item_weight)
else:
    return np.nansum(errors)
```
This looks less elegant to me.

Also, if the weights are uniform, I suspect that `np.nansum(errors * horizon_weight, axis=1)` is less numerically stable than `np.nanmean(errors, axis=1)` (especially if `prediction_length` is very large).

File: timeseries/src/autogluon/timeseries/metrics/abstract.py [hunk 4]
```diff
@@ -92,6 +96,8 @@ def __call__(
                     data_future=data_future,
                     predictions=predictions,
                     target=target,
+                    prediction_length=prediction_length,
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/metrics/abstract.py:98]: Before this PR, a scorer object could be used with arbitrary `prediction_length` 
```python
scorer = check_get_evaluation_metric(...)
# this would work
scorer(data, predictions_for_5, prediction_length=5)
scorer(data, predictions_for_10, prediction_length=10)
```
With the new design, if `horizon_weight is not None`, the scorer will become restricted to only work with a certain `prediction_length = len(horizon_weight)`.

I agree that it makes sense to make the `prediction_length` an attribute of the scorer as well. This should not affect the public API, so we should be fine.

File: timeseries/src/autogluon/timeseries/metrics/__init__.py [hunk 5]
```diff
@@ -75,3 +77,29 @@ def check_get_evaluation_metric(
...
         )
     return scorer
+
+
+@overload
+def check_get_horizon_weight(horizon_weight: None, prediction_length: int) -> None: ...
+@overload
+def check_get_horizon_weight(horizon_weight: list[float], prediction_length: int) -> np.ndarray: ...
+
+def check_get_horizon_weight(horizon_weight: list[float] | None, prediction_length: int) -> Optional[np.ndarray]:
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/metrics/__init__.py:87]: Now that we are making `prediction_length` an attribute of the scorer, we can fold this method under the scorer as well.

File: timeseries/src/autogluon/timeseries/predictor.py [hunk 6]
```diff
@@ -187,15 +196,15 @@ def __init__(
             if std_freq != str(self.freq):
                 logger.info(f"Frequency '{self.freq}' stored as '{std_freq}'")
             self.freq = std_freq
-        self.eval_metric = check_get_evaluation_metric(eval_metric)
-        self.eval_metric_seasonal_period = eval_metric_seasonal_period
+        self.eval_metric: TimeSeriesScorer = check_get_evaluation_metric(eval_metric)
+        self.eval_metric.seasonal_period = eval_metric_seasonal_period
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/predictor.py:200]: This will require making `check_get_evaluation_metric` also taking `prediction_length` as an argument. This makes sense in light of the redesign discussed above.

File: timeseries/src/autogluon/timeseries/predictor.py (hunk 1 above)
Comment (shchur) [reply timeseries/src/autogluon/timeseries/predictor.py:1513]: Good point

File: timeseries/src/autogluon/timeseries/metrics/abstract.py [hunk 7]
```diff
@@ -92,6 +96,8 @@ def __call__(
...
+                    horizon_weight=self.horizon_weight,
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/metrics/abstract.py:99]: There is a potential breaking API change weirdness related to this. Let me address this in another comment after implementing all other changes discussed so far.

**Review by abdulfatir — COMMENTED** [review]
Thanks @shchur! Overall looks great to me. Left some comments.

File: timeseries/src/autogluon/timeseries/metrics/abstract.py [hunk 8]
```diff
@@ -66,18 +77,16 @@ def __call__(
         self,
         data: TimeSeriesDataFrame,
         predictions: TimeSeriesDataFrame,
-        prediction_length: int = 1,
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/metrics/abstract.py:69]: I've added a backward-compatible option to pass the `prediction_length` here.

**Review by abdulfatir — APPROVED** [review]
🚀

File: timeseries/src/autogluon/timeseries/predictor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/predictor.py:96]: Sorry for being late to the party. I understand it's similar in spirit to how `sample_weight` works in most stats software. LGTM.

[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/predictor.py:96]: Do you mean that `sample_weight` usually adds up to 1, or that it usually adds up to `len(y)`? In `sklearn` it seems to be latter (https://scikit-learn.org/stable/modules/generated/sklearn.utils.class_weight.compute_sample_weight.html).

[the remark this answers is not in the document]
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/predictor.py:96]: I meant the latter.

**Review by canerturkmen — APPROVED** [review]
LGTM!
